### ISSUE TITLE
Wrong (maximum) poller-time reported, when poller-process dies

### ISSUE THREAD

**Issue by bernisys** [Cacti/cacti#4219]
# Describe the bug

Despite the relation to spine, i file it here, because i believe this should be reflected also in the cacti main part.
In our setup, spine 1.2.16 is currently crashing a lot for yet unknown reasons, as mentioned in spine issue https://github.com/Cacti/spine/issues/199
This always causes the reported poll time to show a peak value for the given interval.

As far as i can see, it is probably because the end-time is not set when a spine process crashes, but in fact the actual runtime was much lower. And as a result you will see misleading peak times in poll time graphs, if you have them added.

## To Reproduce
Simulate a crash: Let a reasonably long polling cycle run for a while, then hard-kill the processes before it ends naturally.
Check polling time for this interval, it should show >=298 seconds even if the crash terminated the process earlier.

## Expected behavior
Even if the polling cycle is not completed due to the spine crashes, it still ended at some reasonable time. Hence reporting max time is misleading and it would be more helpful to still see the actual runtime of the processes and not a flat line at max polling time. This would in fact reduce the noise level a lot.
We always thought it was really the poll time that's our problem, but in fact it is now pointing to some different problem, probably more related to the central DB connection. Not too easy to see in the logs though, because we have a lot of noise in those, which we currently try to get rid of.

**Comment by TheWitness** [comment]
Just update to the 1.2.x version of spine.

**Comment by TheWitness** [comment]
There have been so many enhancements to spine in 1.2.17 that it really deserves a bigger bump, but we'll leave it at 1.2.17.

**Comment by bernisys** [comment]
Sounds reasonable. From what I've seen in the upgrade-code the table format is not altered significantly, so we have compiled it and use it already in test in combination with cacti 1.2.16 for a few days now without issues. But there's of course far too little load in that environment to see significant changes in the performance. Crashes seem to happen only in high load situations.

We have just updated out prod from 1.2.14 to 1.2.16 this morning and I first want to observe the performance for a couple of hours ... days, then i will push the 1.2.x spine into the prod environment as well, to see if that makes a significant difference.


I also just had another idea, which is related to the hold-off for the poller-start that i proposed in another ticket.
I've seen that the polling itself doesn't take that much time, less than a minute on even our highest loaded poller. From the logs i added i can see that most of the time is actually spent in the postprocessing part (multi-output field splitting and hex conversion etc). This usually ends with lots of data being pushed at the very end of the cycle and this happens on three of our pollers, which means in total there's a huge load of data-processing at the same time. What I now want to try is shifting the start time in the pollers with highest device counts by half a minute each and see if that relaxes the punctual load on the central DB a bit. This could also stop the crashes which we are currently seeing, as they seem to be strongly related to the DB load. (largest machine gone -> time peaking gone)

Will come back with the results. If this works, this can also be of interest for people like @bmfmancini with larger installations to distribute their DB load more evenly.

If this helps, then maybe you should consider to move the data-postprocessing into a background-task that can run during the whole polling cycle instead of doing that on block after the data collection. Once data arrives locally, it can be processed already with one thread in parallel to the polling, and when the polling is done, more processes can be spawned to increase the speed of the postprocessing.

**Comment by bernisys** [comment]
Update: The 1.2.x spine code indeed helped a lot here. The update to Cacti 1.2.16 code alone was initially increasing the stress on the DB and drove the poller time even higher (longer periods of peaking on more pollers). But with the new spine this not only recovered but actually got a lot better. I can definitely recommend the latest spine. As far as i can see there are a lot less crashes than before, but still they are not completely gone yet - will investigate and keep you updated.

In the first hours I just saw an increased amount of row-reads on the central database as well as a much higher query load (questions per second to the DB). Row reads in the main DB increased from 250k up to 1M-1.5M/s and decreased a bit again to steady 800k now. And the queries/s went up from about 500/s to around 1300/s. But after a few hours this has as well settled to the about usual value now - not sure what caused the sudden increase and decrease in the end, but well .. Computers ... they help us solving problems which we wouldn't have if they didn't exist ...

Here are a few impressions from the situation yesterday:

times still hit the ceiling, but much less than before:
![image](https://user-images.githubusercontent.com/32120318/114799066-17cf0d80-9d97-11eb-8f45-21e4aa76929e.png)

DB load went up significantly, but fortunately no rise in slow queries (that's mostly the remote-inserts to poller_output_boost)
![image](https://user-images.githubusercontent.com/32120318/114798920-ca52a080-9d96-11eb-8456-a14e3bd3981a.png)

Boost stats look fine, peaks not constantly rising = good sign
![image](https://user-images.githubusercontent.com/32120318/114798938-d50d3580-9d96-11eb-97a5-13187d418125.png)

Row activities are back to normal, no idea why the row-read went up from 100 to >1k temporarily
![image](https://user-images.githubusercontent.com/32120318/114798947-d9d1e980-9d96-11eb-8101-02f9373cac0d.png)

**Comment by TheWitness** [comment]
Where is the graph showing max_connections, max_used_connections, and Threads_connected?

**Comment by bernisys** [comment]
It's right here:

![image](https://user-images.githubusercontent.com/32120318/114866414-cf920880-9df3-11eb-8a30-d1bc3b7ffff5.png)

FYI: We're resetting the stats by a hourly cron currently, so the max parallel connections look a bit weird.
Want to fix that later, giving the cacti DB user permissions to reset those stats, then this can be handled by the poll script directly. But as you can imagine, i have other prios on my list at the moment.

**Comment by TheWitness** [comment]
So, for me, a time series graph showing those three variables would be awesome.  I'm not sure what Max Parallel Connections is though.

**Comment by bmfmancini** [comment]
Hey that template is cool can I have a copy?

**Comment by bernisys** [comment]
@TheWitness 
"Max Parallel Connections" is a value that increases over time and represents the all-time maximum number of connections that were open at the same time to the DB. So this shows basically the all-time peak value and it must be reset "manually" (cron) using the "flush status;" command. I found it useful and took it in because the "connections" value (counter type) only showed an averaged number over time avter being processed in rrdtool, hence peaks would vanish and the situation might look fine even when you actually have a problem with recurring peaks. With MaxConn you see the actual peak.

@bmfmancini
Sure, I can put it up on one of my repos so you can clone it from there.
Just be patient, have to prepare everything and clean the files.
